stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +112,331 @@ internal func _tryNormalize(
+    UnsafeBufferPointer<UInt16>(rebasing: icuOutputBuffer[..<normalized]),
+    into: outputBuffer
+  ) else {
+    return _foreignReallocateBuffers(
+      readIndex: readIndex,
+      endIndex: endIndex,
+      guts: guts,
+      outputBuffer: &outputBuffer,
+      icuInputBuffer: &icuInputBuffer,
+      icuOutputBuffer: &icuOutputBuffer
+    )
+  }
+
+  return NormalizationResult(
+    amountFilled: transcoded, nextReadPosition: nextIndex, reallocatedBuffers: false
+  )
+}
+
+internal func _foreignReallocateBuffers(
+  readIndex: String.Index,
+  endIndex: String.Index,
+  guts: _StringGuts,
+  outputBuffer: inout UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: inout UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: inout UnsafeMutableBufferPointer<UInt16>
+) -> NormalizationResult {
+  let sourceCount = endIndex.encodedOffset - readIndex.encodedOffset
+  _reallocateBuffers(
+    sourceCount: sourceCount,
+    outputBuffer: &outputBuffer,
+    icuInputBuffer: &icuInputBuffer,
+    icuOutputBuffer: &icuOutputBuffer
+  )
+  var result = _foreignNormalize(
+    readIndex: readIndex,
+    endIndex: endIndex,
+    guts: guts,
+    outputBuffer: &outputBuffer,
+    icuInputBuffer: &icuInputBuffer,
+    icuOutputBuffer: &icuOutputBuffer
+  )
+  result.reallocatedBuffers = true
+  return result
+}
+
+internal func _fastReallocateBuffers(
+  readIndex: String.Index,
+  sourceBuffer: UnsafeBufferPointer<UInt8>,
+  outputBuffer: inout UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: inout UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: inout UnsafeMutableBufferPointer<UInt16>
+) -> NormalizationResult {
+  let sourceCount = sourceBuffer.count - readIndex.encodedOffset
+  _reallocateBuffers(
+    sourceCount: sourceCount,
+    outputBuffer: &outputBuffer,
+    icuInputBuffer: &icuInputBuffer,
+    icuOutputBuffer: &icuOutputBuffer
+  )
+  var result = _fastNormalize(
+    readIndex: readIndex,
+    sourceBuffer: sourceBuffer,
+    outputBuffer: &outputBuffer,
+    icuInputBuffer: &icuInputBuffer,
+    icuOutputBuffer: &icuOutputBuffer
+  )
+  result.reallocatedBuffers = true
+  return result
+}
+
+private func foreignFill(
+  readIndex: String.Index,
+  endIndex: String.Index,
+  _ guts: _StringGuts,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> (Int, Int)? {
+  var index = readIndex
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let cachedEndIndex = endIndex
+  while index != cachedEndIndex {
+    let (scalar, length) = guts.errorCorrectedScalar(startingAt: index.encodedOffset)
+    if scalar._hasNormalizationBoundaryBefore && index != readIndex {
+      break
+    }
+    
+    index = index.encoded(offsetBy: length)
+    
+    for cu in scalar.utf16 {
+      if writeIndex < outputCount {
+        outputBuffer[writeIndex] = cu
+        writeIndex += 1
+      } else {
+        return nil
+      }
+    }
+  }
+  return (index.encodedOffset - readIndex.encodedOffset, writeIndex)
```

Comment:
If taking a `String.Index` streamlines code elsewhere, then we should assert no transcoding. Otherwise, no need to take the index.